plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java

```diff
@@ -0,0 +1,158 @@
+      if (clientId == null) {
+        // Use the current time in milliseconds as a client ID and store it back to the preferences.
+        clientId = UUID.randomUUID().toString();
+        prefs.put(PREF_KEY_CLIENT_ID, clientId);
+      }
+    }
+    return clientId;
+  }
+
+  private static boolean hasUserOptedIn() {
+    IEclipsePreferences prefs = ConfigurationScope.INSTANCE.getNode(ANALYTICS_PREFERENCE_QUALIFIER);
+    if (prefs != null) {
+      return prefs.getBoolean(PREF_KEY_USER_CONSENT, false);
+    }
+    return false;
+  }
+
+  public static void sendPing(String eventType, String eventName,
+      String metadataKey, String metadataValue) {
+    Map<String, String> parametersMap = new HashMap<>(STANDARD_PARAMETERS);
+
+    parametersMap.put("cid", getAnonymizedClientId());
+    parametersMap.put("cd19", eventType);
+    parametersMap.put("cd20", eventName);
+
+    String virtualPageUrl = "/virtual/" + APPLICATION_NAME + "/" + eventType + "/" + eventName;
+    parametersMap.put("dp", virtualPageUrl);
+
+    if (metadataKey != null) {
+      // Event metadata are passed as a (virtual) page title.
+      String virtualPageTitle = metadataKey + "=";
+      if (metadataValue != null) {
+        virtualPageTitle += metadataValue;
+      } else {
+        virtualPageTitle += "null";
+      }
+
+      parametersMap.put("dt", virtualPageTitle);
+    }
+
+    sendPostRequest(parametersMap);
+  }
+
+  private static void sendPostRequest(Map<String, String> parametersMap) {
+    if (Platform.inDevelopmentMode() || !isTrackingIdDefined() || !hasUserOptedIn()) {
+      return;
+    }
+
+    String parametersString = getParametersString(parametersMap);
+    try {
+      URL url = new URL(ANALYTICS_COLLECTION_URL);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setDoOutput(true);
+      connection.setRequestMethod("POST");
+      connection.setRequestProperty("Content-Length", Integer.toString(parametersString.length()));
+
+      try (
+        Writer writer = new OutputStreamWriter(connection.getOutputStream(), "UTF-8")
+      ) {
+        writer.write(parametersString);
+        writer.flush();
+      } finally {
+        connection.disconnect();
+      }
+    } catch (IOException ex) {
+      // Don't try to recover or retry.
+    }
+  }
+
+  @VisibleForTesting
+  static String getParametersString(Map<String, String> parametersMap) {
+    StringBuilder resultBuilder = new StringBuilder();
+    boolean ampersandNeeded = false;
+    for (Map.Entry<String, String> entry : parametersMap.entrySet()) {
+      if (ampersandNeeded) {
+        resultBuilder.append('&');
+      } else {
+        ampersandNeeded = true;
+      }
+      resultBuilder.append(entry.getKey());
+      resultBuilder.append('=');
+      try {
+        resultBuilder.append(URLEncoder.encode(entry.getValue(), "UTF-8"));
+      } catch (UnsupportedEncodingException uee) {
+        throw new AssertionError("UTF-8 not supported");
```

Comment:
I used RuntimeException since the Analytics protocol requires UTF-8 and it seems highly unlikely that any VM does not support UTF-8 these days.